design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,213 @@
+identified in this case are on performance between different types of C-level
+operations, whether certain hardware-level atomic operations are possible, and
+general familiarity of programmers with the representation.
+
+For this particular type we are estimating that the use of ``_BitInt`` types
+will not be such that operations on these types are performance critical.
+
+There seem to be two different regimes for these types.  The "small" regime
+where bit-precise types could be stored in a single register, and the "large"
+regime where bit-precise types must span multiple registers.
+
+Alignment and sizes
+-------------------
+
+These types must be at least byte-aligned so they are addressable, and at least
+rounded to a byte boundary in size for ``sizeof``.  Since these types have an
+aesthetic similarity to bit-fields one might expect better packing in an array
+of ``_BitInt(24)`` than an array of ``int32_t`` types (i.e. packing as good as a
+byte-array).  However, this would require a low alignment of such types and that
+would mean loading and storing of even "small" sized ``_BitInt``'s crossing
+cache boundaries -- leading to an unnecessary performance hit and hindering any
+atomic operations on these.
+
+Hence for "small" sizes we are choosing to define a ``_BitInt(N)`` size and
+alignment according to the smallest Fundamental Data Type which has a bit-size
+greater or equal to ``N``.  Similar for ``unsigned`` versions.
+
+For "large" sizes the only approach considered has been to treat these
+bit-precise types as an array of ``M`` sized chunks, for some ``M``.  The two
+"reasonable" choices for this ``M`` seem to either be register sized or
+double-register sized.  Choosing a register sized chunk would mean smaller sizes
+of types for half of the values of ``N``, while choosing a double-register sized
+chunk would allow atomic operations on types in the range between the register
+and double-register sizes due to the associated extra alignment allowing
+operations like ``CASP`` on aarch64 and ``LDRD`` on aarch32.  Moreover, the
+majority of "large" size use-cases proposed so far are of power-of-two sizes
+like sha256 which would not be in the range which suffers in space-cost from
+this choice.  Finally, defining the ``_BitInt`` representation in this manner
+means that on AArch32 a ``_BitInt(64)`` has the same alignment and size as a
+``int64_t`` which is the largest size defined on that platform, and on AArch64
+a ``_BitInt(128)`` has the same alignment and size as a ``__int128`` which is
+the largest type defined on that platform.  This falls out of the fact that
+double-register size maps to the largest integral Fundamental Data Type defined
+on both platforms.
+
+Hence for "large" sizes we are choosing to define a ``_BitInt(N)`` size and
+alignment by treating them "as if" they are an array of double-register sized
+Fundamental Data Types.
+
+Representation in bits
+----------------------
```

Comment:
Should the above title be "Representation in registers" instead of "...
in bits"?